sql-plugin/src/main/scala/com/nvidia/spark/rapids/ParquetCachedBatchSerializer.scala

```diff
@@ -1056,61 +1073,79 @@ protected class ParquetCachedBatchSerializer extends GpuCachedBatchSerializer {
 
       override def next(): CachedBatch = {
         if (queue.isEmpty) {
-          // to store a row if we have read it but there is no room in the parquet file to put it
-          // we will put it in the next CachedBatch
-          var leftOverRow: Option[InternalRow] = None
-          val rowIterator = getIterator
-          while (rowIterator.hasNext || leftOverRow.nonEmpty) {
-            // Each partition will be a single parquet file
-            var rows = 0
-            // at least a single block
-            val stream = new ByteArrayOutputStream(ByteArrayOutputFile.BLOCK_SIZE)
-            val outputFile: OutputFile = new ByteArrayOutputFile(stream)
-            conf.setConfString(SparkShimImpl.parquetRebaseWriteKey,
-              LegacyBehaviorPolicy.CORRECTED.toString)
-            val recordWriter = SQLConf.withExistingConf(conf) {
+          queue ++= convertRowIteratorToCachedBatch(getIterator)
+        }
+        if (!queue.isEmpty) {
+          queue.dequeue()
+        } else {
+          throw new IllegalStateException("Encountered an empty batch")
+        }
+      }
+
+      protected def convertRowIteratorToCachedBatch(
+          rowIterator: Iterator[InternalRow]): mutable.Queue[CachedBatch] = {
+        val queue = new mutable.Queue[CachedBatch]()
+        // to store a row if we have read it but there is no room in the parquet file to put it
+        // we will put it in the next CachedBatch
+        var leftOverRow: Option[InternalRow] = None
+        while (rowIterator.hasNext || leftOverRow.nonEmpty) {
+          // Each partition will be a single parquet file
+          var rows = 0
+          // at least a single block
+          val stream = new ByteArrayOutputStream(ByteArrayOutputFile.BLOCK_SIZE)
+          val outputFile: OutputFile = new ByteArrayOutputFile(stream)
+          conf.setConfString(SparkShimImpl.parquetRebaseWriteKey,
+            LegacyBehaviorPolicy.CORRECTED.toString)
+          val recordWriter = try {
+            SQLConf.withExistingConf(conf) {
               parquetOutputFileFormat.getRecordWriter(outputFile, hadoopConf)
             }
-            var totalSize = 0
-            while ((rowIterator.hasNext || leftOverRow.nonEmpty)
-                && totalSize < bytesAllowedPerBatch) {
-
-              val row = if (leftOverRow.nonEmpty) {
-                val a = leftOverRow.get
-                leftOverRow = None // reset value
-                a
-              } else {
-                rowIterator.next()
+          } catch {
+            case _: InvalidSchemaException =>
+              // The schema is invalid, most probably it is because of the edge case where there
+              // are no columns in the Dataframe but has rows so let's create an empty CachedBatch
+              queue += new ParquetCachedBatch(rowIterator.size, new Array[Byte](0))
```

Comment:
I am not sure that `rowIterator.size` is what we want.  If you had a clean separate path I would feel a lot better about this, because I cannot tell from the code that this exception will be thrown before anything is read from the iterator, and I cannot tell if leftOverRow is guaranteed to never be set.

Also what happens if we have an InvalidSchema that is not because the schema was empty?